File: go/object.go
package main

import (
	"fmt"
	"sort"
	"time"
)

type Subject struct {
	Name               string
	Code               string
	ProfessorID        int
	EnrolledStudentIDs map[int]bool
	AssignmentGrades   map[int]map[string]float64
	ReportGrades       map[int]map[string]float64
	Assignments        []string
	Reports            []string
}

func NewSubject(name, code string, professorID int) *Subject {
	return &Subject{
		Name:               name,
		Code:               code,
		ProfessorID:        professorID,
		EnrolledStudentIDs: make(map[int]bool),
		AssignmentGrades:   make(map[int]map[string]float64),
		ReportGrades:       make(map[int]map[string]float64),
		Assignments:        make([]string, 0),
		Reports:            make([]string, 0),
	}
}

func (s *Subject) EnrollStudent(studentID int) {
	s.EnrolledStudentIDs[studentID] = true
}

func (s *Subject) IsStudentEnrolled(studentID int) bool {
	return s.EnrolledStudentIDs[studentID]
}

func (s *Subject) AddAssignment(assignmentName string) {
	s.Assignments = append(s.Assignments, assignmentName)
}

func (s *Subject) AddReport(reportName string) {
	s.Reports = append(s.Reports, reportName)
}

func (s *Subject) RemoveReport(reportName string) bool {
	for i, r := range s.Reports {
		if r == reportName {
			s.Reports = append(s.Reports[:i], s.Reports[i+1:]...)
			return true
		}
	}
	return false
}

func (s *Subject) GradeAssignment(studentID int, assignmentName string, grade float64) {
	if s.IsStudentEnrolled(studentID) && s.HasAssignment(assignmentName) {
		if s.AssignmentGrades[studentID] == nil {
			s.AssignmentGrades[studentID] = make(map[string]float64)
		}
		s.AssignmentGrades[studentID][assignmentName] = grade
	}
}

func (s *Subject) GradeReport(studentID int, reportName string, grade float64) {
	if s.IsStudentEnrolled(studentID) {
		if s.ReportGrades[studentID] == nil {
			s.ReportGrades[studentID] = make(map[string]float64)
		}
		s.ReportGrades[studentID][reportName] = grade
	}
}

func (s *Subject) GradeAllReports(reportName string, grade float64, participants []int) {
	if !s.HasReport(reportName) {
		return
	}

	if len(participants) == 0 {
		for studentID := range s.EnrolledStudentIDs {
			if s.ReportGrades[studentID] == nil {
				s.ReportGrades[studentID] = make(map[string]float64)
			}
			s.ReportGrades[studentID][reportName] = grade
		}
	} else {
		for _, studentID := range participants {
			if s.IsStudentEnrolled(studentID) {
				if s.ReportGrades[studentID] == nil {
					s.ReportGrades[studentID] = make(map[string]float64)
				}
				s.ReportGrades[studentID][reportName] = grade
			}
		}
	}
}

func (s *Subject) GetStudentAssignmentGrade(studentID int, assignmentName string) float64 {
	if grades, ok := s.AssignmentGrades[studentID]; ok {
		if grade, ok := grades[assignmentName]; ok {
			return grade
		}
	}
	return -1.0
}

func (s *Subject) GetStudentReportGrade(studentID int, reportName string) float64 {
	if grades, ok := s.ReportGrades[studentID]; ok {
		if grade, ok := grades[reportName]; ok {
			return grade
		}
	}
	return -1.0
}

func (s *Subject) GetEnrolledStudents() []int {
	students := make([]int, 0, len(s.EnrolledStudentIDs))
	for id := range s.EnrolledStudentIDs {
		students = append(students, id)
	}
	sort.Ints(students)
	return students
}

func (s *Subject) IsProfessor(professorID int) bool {
	return s.ProfessorID == professorID
}

func (s *Subject) HasAssignment(assignmentName string) bool {
	for _, a := range s.Assignments {
		if a == assignmentName {
			return true
		}
	}
	return false
}

func (s *Subject) HasReport(reportName string) bool {
	for _, r := range s.Reports {
		if r == reportName {
			return true
		}
	}
	return false
}

func (s *Subject) GetAssignments() []string {
	return s.Assignments
}

func (s *Subject) GetReports() []string {
	return s.Reports
}

func (s *Subject) GenerateFinalReport(studentNames map[int]string) {
	fmt.Printf("\n=== ИТОГОВЫЙ ОТЧЕТ: %s (%s) ===\n", s.Name, s.Code)
	fmt.Printf("ID преподавателя: %d\n", s.ProfessorID)
	fmt.Printf("Зачисленных студентов: %d\n", len(s.EnrolledStudentIDs))
	fmt.Println("==============================================")

	students := s.GetEnrolledStudents()
	for _, studentID := range students {
		studentName := studentNames[studentID]
		if studentName == "" {
			studentName = "Неизвестный"
		}

		fmt.Printf("\nСтудент: %s (ID: %d)\n", studentName, studentID)

		total := 0.0
		count := 0

		if assignmentGrades, ok := s.AssignmentGrades[studentID]; ok {
			fmt.Println("Оценки за задания:")
			// Сортируем названия заданий для красивого вывода
			var assignments []string
			for assignment := range assignmentGrades {
				assignments = append(assignments, assignment)
			}
			sort.Strings(assignments)

			for _, assignment := range assignments {
				grade := assignmentGrades[assignment]
				fmt.Printf("  %-20s: %6.2f\n", assignment, grade)
				total += grade
				count++
			}
		}

		if reportGrades, ok := s.ReportGrades[studentID]; ok {
			fmt.Println("Оценки за доклады:")
			// Сортируем названия докладов
			var reports []string
			for report := range reportGrades {
				reports = append(reports, report)
			}
			sort.Strings(reports)

			for _, report := range reports {
				grade := reportGrades[report]
				fmt.Printf("  %-20s: %6.2f\n", report, grade)
				total += grade
				count++
			}
		}

		if count > 0 {
			fmt.Printf("  Среднее: %.2f\n", total/float64(count))
			fmt.Printf("  Суммарное: %.2f\n", total)
		} else {
			fmt.Println("  Нет оценок")
		}
	}
	fmt.Println()
}

func (s *Subject) SetAssignmentGrades(grades map[int]map[string]float64) {
	s.AssignmentGrades = grades
}

// SetReportGrades устанавливает оценки за доклады
func (s *Subject) SetReportGrades(grades map[int]map[string]float64) {
	s.ReportGrades = grades
}

// GetAllAssignmentGrades возвращает все оценки за задания
func (s *Subject) GetAllAssignmentGrades() map[int]map[string]float64 {
	return s.AssignmentGrades
}

// GetAllReportGrades возвращает все оценки за доклады
func (s *Subject) GetAllReportGrades() map[int]map[string]float64 {
	return s.ReportGrades
}

// GetEnrolledStudentIds возвращает ID всех зачисленных студентов
func (s *Subject) GetEnrolledStudentIds() []int {
	students := make([]int, 0, len(s.EnrolledStudentIDs))
	for id := range s.EnrolledStudentIDs {
		students = append(students, id)
	}
	sort.Ints(students)
	return students
}

// GetAssignmentList возвращает список всех заданий
func (s *Subject) GetAssignmentList() []string {
	return s.Assignments
}

// GetReportList возвращает список всех докладов
func (s *Subject) GetReportList() []string {
	return s.Reports
}

func (s *Subject) GetStudentGradesSummary(studentID int) []struct {
	Item  string
	Grade float64
} {
	var result []struct {
		Item  string
		Grade float64
	}

	if assignmentGrades, ok := s.AssignmentGrades[studentID]; ok {
		for assignment, grade := range assignmentGrades {
			result = append(result, struct {
				Item  string
				Grade float64
			}{
				Item:  "Задание: " + assignment,
				Grade: grade,
			})
		}
	}

	if reportGrades, ok := s.ReportGrades[studentID]; ok {
		for report, grade := range reportGrades {
			result = append(result, struct {
				Item  string
				Grade float64
			}{
				Item:  "Доклад: " + report,
				Grade: grade,
			})
		}
	}

	return result
}

type Assignment struct {
	Name        string
	SubjectName string
	MaxScore    float64
}

func NewAssignment(name, subjectName string, maxScore float64) *Assignment {
	return &Assignment{
		Name:        name,
		SubjectName: subjectName,
		MaxScore:    maxScore,
	}
}

type Report struct {
	Topic              string
	SubjectName        string
	SignedUpStudentIDs map[int]bool
	Date               time.Time
	MaxParticipants    int
	IsCompleted        bool
}

func NewReport(topic, subjectName string, maxParticipants int) *Report {
	return &Report{
		Topic:              topic,
		SubjectName:        subjectName,
		SignedUpStudentIDs: make(map[int]bool),
		Date:               time.Now(),
		MaxParticipants:    maxParticipants,
		IsCompleted:        false,
	}
}

func (r *Report) AddStudent(studentID int) bool {
	if r.IsFull() || r.IsCompleted {
		return false
	}

	if !r.SignedUpStudentIDs[studentID] {
		r.SignedUpStudentIDs[studentID] = true
		return true
	}
	return false
}

func (r *Report) RemoveStudent(studentID int) bool {
	if r.SignedUpStudentIDs[studentID] {
		delete(r.SignedUpStudentIDs, studentID)
		return true
	}
	return false
}

func (r *Report) IsFull() bool {
	return len(r.SignedUpStudentIDs) >= r.MaxParticipants
}

func (r *Report) HasStudent(studentID int) bool {
	return r.SignedUpStudentIDs[studentID]
}

func (r *Report) GetSignedUpStudents() []int {
	students := make([]int, 0, len(r.SignedUpStudentIDs))
	for id := range r.SignedUpStudentIDs {
		students = append(students, id)
	}
	sort.Ints(students)
	return students
}

func (r *Report) MarkAsCompleted() {
	r.IsCompleted = true
}

File: cpp/object.cpp
#include "object.h"
#include "student.h"

using namespace std;

Subject::Subject(const string& name, const string& code, int professorId)
    : name(name), code(code), professorId(professorId) {}

void Subject::enrollStudent(int studentId) {
    enrolledStudentIds.insert(studentId);
}

bool Subject::isStudentEnrolled(int studentId) const {
    return enrolledStudentIds.find(studentId) != enrolledStudentIds.end();
}

void Subject::addAssignment(const string& assignmentName) {
    assignments.push_back(assignmentName);
}

void Subject::addReport(const string& reportName) {
    reports.push_back(reportName);
}

bool Subject::removeReport(const string& reportName) {
    auto it = find(reports.begin(), reports.end(), reportName);
    if (it != reports.end()) {
        reports.erase(it);
        return true;
    }
    return false;
}

void Subject::gradeAssignment(int studentId, const string& assignmentName, double grade) {
    if (isStudentEnrolled(studentId) && hasAssignment(assignmentName)) {
        assignmentGrades[studentId][assignmentName] = grade;
    }
}

void Subject::gradeReport(int studentId, const string& reportName, double grade) {
    if (isStudentEnrolled(studentId)) {
        reportGrades[studentId][reportName] = grade;
    }
}

void Subject::gradeAllReports(const string& reportName, double grade, const vector<int>& participants) {
    if (!hasReport(reportName)) return;
    
    if (participants.empty()) {
        for (int studentId : enrolledStudentIds) {
            reportGrades[studentId][reportName] = grade;
        }
    } else {
        for (int studentId : participants) {
            if (enrolledStudentIds.find(studentId) != enrolledStudentIds.end()) {
                reportGrades[studentId][reportName] = grade;
            }
        }
    }
}

double Subject::getStudentAssignmentGrade(int studentId, const string& assignmentName) const {
    auto itStudent = assignmentGrades.find(studentId);
    if (itStudent != assignmentGrades.end()) {
        auto itAssignment = itStudent->second.find(assignmentName);
        if (itAssignment != itStudent->second.end()) {
            return itAssignment->second;
        }
    }
    return -1.0;
}

double Subject::getStudentReportGrade(int studentId, const string& reportName) const {
    auto itStudent = reportGrades.find(studentId);
    if (itStudent != reportGrades.end()) {
        auto itReport = itStudent->second.find(reportName);
        if (itReport != itStudent->second.end()) {
            return itReport->second;
        }
    }
    return -1.0;
}

vector<int> Subject::getEnrolledStudents() const {
    return vector<int>(enrolledStudentIds.begin(), enrolledStudentIds.end());
}

vector<string> Subject::getAssignments() const {
    return assignments;
}

vector<string> Subject::getReports() const {
    return reports;
}

void Subject::generateFinalReport(const map<int, string>& studentNames) const {
    cout << "\n=== ИТОГОВЫЙ ОТЧЕТ: " << name << " (" << code << ") ===\n";
    cout << "ID преподавателя: " << professorId << "\n";
    cout << "Зачисленных студентов: " << enrolledStudentIds.size() << "\n";
    cout << "==============================================\n";
    
    for (int studentId : enrolledStudentIds) {
        auto itName = studentNames.find(studentId);
        string studentName = (itName != studentNames.end()) ? itName->second : "Неизвестный";
        
        cout << "\nСтудент: " << studentName << " (ID: " << studentId << ")\n";
        
        double total = 0.0;
        int count = 0;
        
        auto itAssignments = assignmentGrades.find(studentId);
        if (itAssignments != assignmentGrades.end()) {
            cout << "Оценки за задания:\n";
            for (const auto& [assignment, grade] : itAssignments->second) {
                cout << "  " << left << setw(20) << assignment 
                          << ": " << right << setw(6) << fixed 
                          << setprecision(2) << grade << "\n";
                total += grade;
                count++;
            }
        }
        
        auto itReports = reportGrades.find(studentId);
        if (itReports != reportGrades.end()) {
            cout << "Оценки за доклады:\n";
            for (const auto& [report, grade] : itReports->second) {
                cout << "  " << left << setw(20) << report 
                          << ": " << right << setw(6) << fixed 
                          << setprecision(2) << grade << "\n";
                total += grade;
                count++;
            }
        }
        
        if (count > 0) {
            cout << "  Среднее: " << fixed << setprecision(2) << (total / count) << "\n";
            cout << "  Суммарное: " << fixed << setprecision(2) << total << "\n";
        } else {
            cout << "  Нет оценок\n";
        }
    }
    cout << "\n";
}

bool Subject::hasAssignment(const string& assignmentName) const {
    return find(assignments.begin(), assignments.end(), assignmentName) != assignments.end();
}

bool Subject::hasReport(const string& reportName) const {
    return find(reports.begin(), reports.end(), reportName) != reports.end();
}

vector<pair<string, double>> Subject::getStudentGradesSummary(int studentId) const {
    vector<pair<string, double>> result;
    
    auto itAssignments = assignmentGrades.find(studentId);
    if (itAssignments != assignmentGrades.end()) {
        for (const auto& [assignment, grade] : itAssignments->second) {
            result.emplace_back("Задание: " + assignment, grade);
        }
    }
    
    auto itReports = reportGrades.find(studentId);
    if (itReports != reportGrades.end()) {
        for (const auto& [report, grade] : itReports->second) {
            result.emplace_back("Доклад: " + report, grade);
        }
    }
    
    return result;
}

Report::Report(const string& topic, const string& subjectName, int maxParticipants)
    : topic(topic), subjectName(subjectName), maxParticipants(maxParticipants), 
      isCompleted(false) {
    date = time(nullptr);
}

bool Report::addStudent(int studentId) {
    if (isFull() || isCompleted) {
        return false;
    }
    
    if (signedUpStudentIds.find(studentId) == signedUpStudentIds.end()) {
        signedUpStudentIds.insert(studentId);
        return true;
    }
    
    return false;
}

bool Report::removeStudent(int studentId) {
    return signedUpStudentIds.erase(studentId) > 0;
}

bool Report::isFull() const {
    return signedUpStudentIds.size() >= maxParticipants;
}

bool Report::hasStudent(int studentId) const {
    return signedUpStudentIds.find(studentId) != signedUpStudentIds.end();
}

vector<int> Report::getSignedUpStudents() const {
    return vector<int>(signedUpStudentIds.begin(), signedUpStudentIds.end());
}